src/tracks/playabletrack/wavetrack/ui/WaveChannelView.cpp

```diff
@@ -1524,72 +1524,68 @@ std::shared_ptr<ChannelVRulerControls> WaveChannelView::DoGetVRulerControls()
 
 namespace
 {
-   // Returns an offset in seconds to be applied to the right clip 
-   // boundary so that it does not overlap the last sample
-   double CalculateAdjustmentForZoomLevel(
-      const wxRect& viewRect, 
-      const ZoomInfo& zoomInfo, 
-      int rate, 
-      double& outAveragePPS,
-      //Is zoom level sufficient to show individual samples?
-      bool& outShowSamples)
-   {
-      static constexpr double pixelsOffset{ 2 };//The desired offset in pixels
+// Returns an offset in seconds to be applied to the right clip
+// boundary so that it does not overlap the last sample
+double CalculateAdjustmentForZoomLevel(double avgPixPerSecond, bool showSamples)
+{
+   constexpr double pixelsOffset { 2 }; // The desired offset in pixels
+   if (showSamples)
+      // adjustment so that the last circular point doesn't appear
+      // to be hanging off the end
+      return pixelsOffset /
+             avgPixPerSecond; // pixels / ( pixels / second ) = seconds
+   return .0;
+}
 
-      auto h = zoomInfo.PositionToTime(0, 0, true);
-      auto h1 = zoomInfo.PositionToTime(viewRect.width, 0, true);
+double GetBlankSpaceBeforePlayEndTime(const WaveClip& clip)
+{
+   return 0.99 * clip.GetStretchRatio() / clip.GetRate();
+}
 
-      // Determine whether we should show individual samples
-      // or draw circular points as well
-      outAveragePPS = viewRect.width / (rate * (h1 - h));// pixels per sample
-      outShowSamples = outAveragePPS > 0.5;
+double GetPixelsPerSecond(const wxRect& viewRect, const ZoomInfo& zoomInfo)
+{
+   const auto h = zoomInfo.PositionToTime(0, 0, true);
+   const auto trackRectT1 = zoomInfo.PositionToTime(viewRect.width, 0, true);
+   return viewRect.width / (trackRectT1 - h);
+}
 
-      if(outShowSamples)
-         // adjustment so that the last circular point doesn't appear
-         // to be hanging off the end
-         return  pixelsOffset / (outAveragePPS * rate); // pixels / ( pixels / second ) = seconds
-      return .0;
-   }
+bool ShowIndividualSamples(
+   const wxRect& viewRect, const ZoomInfo& zoomInfo, int sampleRate,
+   double stretchRatio, double pixelsPerSecond)
+{
+   const auto secondsPerSample = stretchRatio / sampleRate;
+   const auto pixelsPerSample = pixelsPerSecond * secondsPerSample;
+   return pixelsPerSample > 0.5;
+}
 }
 
-ClipParameters::ClipParameters
-   (bool spectrum, const SampleTrack *track, const WaveClip *clip, const wxRect &rect,
-   const SelectedRegion &selectedRegion, const ZoomInfo &zoomInfo)
+ClipParameters::ClipParameters(
+   const SampleTrack* track, const WaveClip* clip, const wxRect& rect,
+   const SelectedRegion& selectedRegion, const ZoomInfo& zoomInfo)
```

Comment:
Arguments `track` and `selectedRegion` are not used

Reply:
How could I miss that ... thank you.